libsignal-protocol/tests/libsignal-protocol-c-tests.rs

```diff
@@ -400,3 +411,124 @@ fn test_optional_one_time_pre_key() {
+    let alice_store = sig::store_context(
+        &ctx,
+        InMemoryPreKeyStore::default(),
+        InMemorySignedPreKeyStore::default(),
+        InMemorySessionStore::default(),
+        InMemoryIdentityKeyStore::new(
+            sig::generate_registration_id(&ctx, 0).unwrap(),
+            &alice_identity,
+        ),
+    )
+    .unwrap();
+    let alice_session_builder =
+        sig::session_builder(&ctx, &alice_store, &bob_address);
+
+    // Create Bob's data store and pre key bundle
+    let registration_id = sig::generate_registration_id(&ctx, 0).unwrap();
+    let bob_identity_key_pair = sig::generate_identity_key_pair(&ctx).unwrap();
+    let bob_store = sig::store_context(
+        &ctx,
+        InMemoryPreKeyStore::default(),
+        InMemorySignedPreKeyStore::default(),
+        InMemorySessionStore::default(),
+        InMemoryIdentityKeyStore::new(registration_id, &bob_identity_key_pair),
+    )
+    .unwrap();
+
+    let bob_local_registration_id = bob_store.registration_id().unwrap();
+    let bob_signed_pre_key_pair = sig::generate_signed_pre_key(
+        &ctx,
+        &bob_identity_key_pair,
+        22,
+        SystemTime::now(),
+    )
+    .unwrap();
+    let bob_signed_pre_key_public_serialized = bob_signed_pre_key_pair
+        .key_pair()
+        .public()
+        .serialize()
+        .unwrap();
+    let bob_signed_pre_key_signature = sig::calculate_signature(
+        &ctx,
+        &bob_identity_key_pair.private(),
+        bob_signed_pre_key_public_serialized.as_slice(),
+    )
+    .unwrap();
+    bob_store
+        .store_signed_pre_key(&bob_signed_pre_key_pair)
+        .unwrap();
+
+    // Generate pre key for bob
+    let mut bob_pre_keys = sig::generate_pre_keys(&ctx, 2, 4).unwrap();
+    let bob_pre_key = bob_pre_keys.next().unwrap();
+    bob_store.store_pre_key(&bob_pre_key).unwrap();
+
+    let bob_pre_key_bundle = PreKeyBundle::builder()
+        .registration_id(bob_local_registration_id)
+        .identity_key(&bob_identity_key_pair.public())
+        .device_id(1)
+        .pre_key(bob_pre_key.id(), &bob_pre_key.key_pair().public())
+        .signed_pre_key(
+            bob_signed_pre_key_pair.id(),
+            &bob_signed_pre_key_pair.key_pair().public(),
+        )
+        .signature(bob_signed_pre_key_signature.as_slice())
+        .build()
+        .unwrap();
+
+    // Have Alice process Bob's pre key bundle
+    alice_session_builder
+        .process_pre_key_bundle(&bob_pre_key_bundle)
+        .unwrap();
+
+    // create alice's session cipher
+    let alice_session_cipher =
+        sig::SessionCipher::new(&ctx, &alice_store, &bob_address).unwrap();
+
+    // Let Alice send a message to bob
+    let msg = "Hello bob!";
+    let outgoing_message =
+        alice_session_cipher.encrypt(msg.as_bytes()).unwrap();
+
+    // Convert to an incoming message (this is technically a downcast from
+    // CiphertextMessage to PreKeySignalMessage)
```

Comment:
Do we have any guarantee that `outgoing_message` will always be `PreKeySignalMessage`? If so, it might be easier if the `SessionCipher::encrypt()` method returns a `PreKeySignalMessage` directly seeing as that's the more specific "class".

Reply:
No, it won't always be `PreKeySignalMessage`. A few lines below in the test a `SignalMessage` is returned.

To be more precise: it will be a `PreKeySignalMessage` until a message reply has been received. This means that if Alice initiates the conversation with Bob she will never receive a `PreKeySignalMessage` as the first message she gets from Bob will be a reply. 
Alice will send `PreKeySignalMessage`s to Bob until she receives a message from Bob. After this, any subsequent messages sent to Bob will be `SignalMessage`s.